### ISSUE TITLE
BT loss value on val set every N training epochs without classifier

### ISSUE THREAD

**Issue by neerajwagh** [facebookresearch/barlowtwins#26]
Hi, 

Thanks for making it painless for others to use and build upon your work.

I'm curious to know what the loss function value looks like during the training process on the val set of images. In my case of using spectrogram images on a resnet-18 backbone, the network trains well but the val loss value is very noisy and shows no clear trend. For example, I'm not sure what to make of this:

![image](https://user-images.githubusercontent.com/17927658/125985888-55284104-32eb-47bc-aa0d-781014441a89.png)

In the original experiments, was the BT loss value on the val set tracked every N epochs (without any classifier involved)? As far as I can tell, the loss reported in the 'val logs' txt file is of the classifier that was trained on top of the embeddings generating using a single training checkpoint.

The loss curve on the val set should be expected to follow typical behavior, right? Any pointers as to what might be causing a noisy val loss?

**Comment by jzbontar** [comment]
We currently don't log the BT loss on the validation set. In the original experiments we trained a linear classifier in parallel with BT training and tracked its train and val top-1 accuracy.

I don't know why the train and val BT loss would differ so much. I can run this on ImageNet and report back.

**Comment by neerajwagh** [comment]
Hi @jzbontar, thanks for clarifying your val setup.

**It'd be great if you could run a quick experiment (possibly on 10% training data, and a smaller embedding size) to check the BT loss value on the val set in parallel with BT training.** When you try this, you may want track_running_stats=False at https://github.com/facebookresearch/barlowtwins/blob/574f589af8a2cb6214e0e537ca768ea7ae31ab19/main.py#L205 (It makes sense for the val BT loss to use batch statistics computed from the val batch instead of using the train batch statistics, right? From what I understand, the last BN layer only exists to make the loss calculation simpler.)

In your view, is it okay for BT (pre-)training to overfit on the train set or should we stop BT pre-training when BT val loss starts diverging?

**Comment by jzbontar** [comment]
I trained BT on ImageNet training set (1.28M images) and computed the BT loss on the ImageNet validation set (50k images) after each training epoch. I used model.train() for both training and validation, because I had some trouble getting `track_running_stats=False` to work. I trained for 40 epochs using the following command:

`python main.py <imagenet_path> --epochs 40 --batch-size 1024 --learning-rate-weights 0.2 --learning-rate-biase 0.005 --weight-decay 1e-06 --lambd 0.005`

In my experiments the validation loss does not show the behavior that you observed, but rather follows the training loss pretty well. See figure below:  

![foo](https://user-images.githubusercontent.com/1426550/126603999-469fb454-0b90-4ae0-9670-ceb249213670.png)

My best guess is that your training set is small and that BT overfits it.

> In your view, is it okay for BT (pre-)training to overfit on the train set or should we stop BT pre-training when BT val loss starts diverging?

I guess it all depends on how you intend to use the representations learnt by BT. If you plan to use it for a downstream task, I would measure the accuracy on that task and stop training based on that.

**Comment by neerajwagh** [comment]
Oh, that looks neat!

I'm aiming for task-agnostic and dataset-agnostic embeddings, so I cannot have a backbone that overfits my training set. Even if the downstream task is fixed to a specific one, I'm still worried the embeddings may not work for the same task done using a different downstream dataset. So I guess I need to continue debugging until I see healthy loss curves.

You're right, my dataset is much smaller: train #samples: 140k, val #samples: 54k. So I'll probably try a smaller projector and sanity check my train and val distributions.

In your eval loop, do you have the same augmentation process as in your training loop?

**Comment by jzbontar** [comment]
> In your val loop for this particular experiment, do you have the same augmentation process as in your training loop?

Yes, the training and validation datasets have the same augmentation process.

**Comment by HareshKarnan** [comment]
@neerajwagh were you able to get the BTLoss to reduce on the validation set ?
